kernel/src/checkpoint/mod.rs

```diff
@@ -0,0 +1,485 @@
+//!   multi-file support, but the current implementation only supports single-file checkpoints.
+//! - Multi-file V1 checkpoints are DEPRECATED.
+//!
+//! ## Example: Writing a classic-named V1/V2 checkpoint (depending on `v2Checkpoints` feature support)
+//!
+//! TODO(seb): unignore example
+//! ```ignore
+//! let path = "./tests/data/app-txn-no-checkpoint";
+//! let engine = Arc::new(SyncEngine::new());
+//! let table = Table::try_from_uri(path)?;
+//!
+//! // Create a checkpoint writer for the table at a specific version
+//! let mut writer = table.checkpoint(&engine, Some(2))?;
+//!
+//! // Retrieve checkpoint data
+//! let checkpoint_data = writer.get_checkpoint_info()?;
+//!
+//! /* Write checkpoint data to file and collect metadata about the write */
+//! /* The implementation of the write is storage-specific and not shown */
+//! /* IMPORTANT: All data must be written before finalizing the checkpoint */
+//!
+//! // Finalize the checkpoint by writing the _last_checkpoint file
+//! writer.finalize_checkpoint(&engine, &checkpoint_metadata)?;
+//! ```
+//!
+//! This module, along with its submodule `checkpoint/log_replay.rs`, provides the full
+//! API and implementation for generating checkpoints. See `checkpoint/log_replay.rs` for details
+//! on how log replay is used to filter and deduplicate actions for checkpoint creation.
+use crate::{
+    actions::{
+        schemas::GetStructField, Add, Metadata, Protocol, Remove, SetTransaction, Sidecar,
+        ADD_NAME, METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME, SET_TRANSACTION_NAME, SIDECAR_NAME,
+    },
+    expressions::Scalar,
+    path::ParsedLogPath,
+    schema::{DataType, SchemaRef, StructField, StructType},
+    snapshot::Snapshot,
+    DeltaResult, Engine, EngineData, Error, EvaluationHandlerExtension,
+};
+use log_replay::{checkpoint_actions_iter, CheckpointData};
+use std::{
+    cell::RefCell,
+    rc::Rc,
+    sync::{Arc, LazyLock},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use url::Url;
+
+mod log_replay;
+#[cfg(test)]
+mod tests;
+
+/// Schema for extracting relevant actions from log files during checkpoint creation
+static CHECKPOINT_READ_SCHEMA: LazyLock<SchemaRef> = LazyLock::new(|| {
+    StructType::new([
+        Option::<Add>::get_struct_field(ADD_NAME),
+        Option::<Remove>::get_struct_field(REMOVE_NAME),
+        Option::<Metadata>::get_struct_field(METADATA_NAME),
+        Option::<Protocol>::get_struct_field(PROTOCOL_NAME),
+        Option::<SetTransaction>::get_struct_field(SET_TRANSACTION_NAME),
+        Option::<Sidecar>::get_struct_field(SIDECAR_NAME),
+    ])
+    .into()
+});
+
+/// Returns the schema for reading Delta log actions during checkpoint creation
+fn get_checkpoint_read_schema() -> &'static SchemaRef {
+    &CHECKPOINT_READ_SCHEMA
+}
+
+/// Contains the path and data for a single-file checkpoint
+#[allow(unused)] // TODO(seb): Make pub for roll-out
+pub(crate) struct SingleFileCheckpointData {
```

Comment:
for now just checkpoint data?
```suggestion
pub(crate) struct CheckpointData {
```

Reply:
todo! Will update once we replace the other `CheckpointData` with the incoming `FilteredEngineData.